src/solvers/z3/z3_conv.cpp

```diff
@@ -1285,29 +1283,38 @@ void z3_convt::dump_smt()
   const std::string &filename = options.get_option("output");
   if(!filename.empty())
   {
+    // print to output file
     std::ofstream out(filename.c_str());
-    if(out)
-    {
-      // Add whatever logic is needed.
-      // Add sovler specific declarations as well.
-      out << "(set-info :smt-lib-version 2.6)\n";
-      out << "(set-option :print-success true)\n";
-      out << "(set-option :produce-models true)\n";
-      out << "; Asserts from ESMBC starts\n";
-      out << solver; // All VCC conditions in SMTLIB format.
-      out << "; Asserts from ESMBC ends\n";
-      out << "(apply (then simplify solve-eqs))\n";
-      out << "(check-sat)\n";
-      out << "(get-model)\n";
-      out << "(exit)\n";
-    }
+    print_smt_formulae<std::ofstream>(out);
+  }
+  else
+  {
+    // print to screen
+    default_message msg;
+    std::ostringstream oss;
+    print_smt_formulae<std::ostringstream>(oss);
+    msg.debug(oss.str());
   }
+}
 
+template <typename T>
+void z3_convt::print_smt_formulae(T &dest)
```

Comment:
I believe you could avoid the template by declaring `dest` to be a reference to `std::ostream` as the template parameters of both instantiations are derived from it.